### PR TITLE
Improve multi-node multi-gpu tutorial

### PR DESCRIPTION
**Pull request by flxmr**
So, in response to my remarks in #8071 I now prepared this PR for updating the multi-node documentation (sadly no student contrib, they can prep multi-gpu metrics).

Reasoning:
- if pyg has a tutorial on DDP it's for people who essentially grow within pyg to this usecase. This should be reflected in telling them more about this and be clear on what is done (like `torch.multiprocessing` injecting the rank. it seems very random otherwise).
- in the same vein, the multi-node tutorial should evolve from the single-node one imho. 
- embedding the pyxis-container as the only way to do things is good for Nvidia, but bad for people who arrive on a system without that.

Things I skipped:
- setting up the worker-count. I would say people should just figure this out on their own. the `os.sched_getaffinity` is cool, but this is a very general problem too and I essentially do it manually now.
- anything GRES-related. I doubt that anyone growing into this and installing slurm on their computers will not grok this. Every multi-user, public research system has GRES I'd say (and everyone has their internal docs).

### CODE REVIEW

File: docs/source/tutorial/multi_node_multi_gpu_vanilla.rst [hunk 1]
```diff
@@ -1,109 +1,112 @@
-Multi-Node-Multi-GPU Training in Pure PyTorch
-=============================================
+Multi-Node-Multi-GPU Training using SLURM
+=========================================================
```
Comment (akihironitta) [inline docs/source/tutorial/multi_node_multi_gpu_vanilla.rst:3]:
```suggestion
Multi-Node-Multi-GPU Training using SLURM
=========================================

```

File: docs/source/tutorial/multi_node_multi_gpu_vanilla.rst [hunk 2]
```diff
@@ -1,109 +1,112 @@
...
 
-    def run(device, world_size, data, model):
-        local_id = int(os.environ['LOCAL_RANK'])
-        rank = torch.distributed.get_rank()
-        torch.cuda.set_device(local_id)
-        device = torch.device(local_id)
-        ...
+    export MASTER_PORT=$(expr 10000 + $(echo -n $SLURM_JOBID | tail -c 4))
+    master_addr=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
+    export MASTER_ADDR=$master_addr
```
Comment (akihironitta) [inline docs/source/tutorial/multi_node_multi_gpu_vanilla.rst:43]:
```suggestion
    export MASTER_ADDR=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
```

File: docs/source/tutorial/multi_node_multi_gpu_vanilla.rst [hunk 3]
```diff
@@ -1,109 +1,112 @@
...
+.. code-block:: console
 
-        loader = NeighborLoader(
-            dataset,
-            input_nodes=input_nodes,
-            num_neighbors=[10, 10],
-            batch_size=128,
-            shuffle=True,
-        )
+    $ sbatch distributed_sampling_multinode.sbatch
```
Comment (akihironitta) [inline docs/source/tutorial/multi_node_multi_gpu_vanilla.rst:60]: nit: Should we remove `$` for consistency?
```suggestion
    sbatch distributed_sampling_multinode.sbatch
```

File: docs/source/tutorial/multi_node_multi_gpu_vanilla.rst [hunk 4]
```diff
@@ -1,109 +1,112 @@
...
-                loss = F.cross_entropy(out, batch.y)
-                loss.backward()
-                optimizer.step()
 
-1. We put our :class:`~torch_geometric.nn.models.GCN` model on its respective :obj:`device` and wrap it inside :class:`~torch.nn.parallel.DistributedDataParallel` while we pass :obj:`local_id` to its :obj:`device_id` parameter.
-2. We then set up our optimizer for training.
-3. We then split our input/seed nodes into :obj:`world_size` many chunks for each GPU, and initialize the :class:`~torch_geometric.loader.NeighborLoader` class to only operate on its specific subset of nodes.
-4. Finally, we iterate over epochs and batches to train our GNN as usual.
+Using a cluster configured with pyxis-containers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (akihironitta) [inline docs/source/tutorial/multi_node_multi_gpu_vanilla.rst:64]:
```suggestion
Using a cluster configured with pyxis-containers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/source/tutorial/multi_node_multi_gpu_vanilla.rst [hunk 5]
```diff
@@ -1,109 +1,112 @@
...
+The container has all necessary environment-variables populated already.
 
-NVIDIA provides a ready-to-use :pyg:`PyG` container that is updated each month with the latest from NVIDIA and :pyg:`PyG`.
 You can sign up for early access `here <https://developer.nvidia.com/pyg-container-early-access>`_.
 General availability on `NVIDIA NGC <https://www.ngc.nvidia.com/>`_ is set for the end of 2023.
 Alternatively, see `docker.com <https://www.docker.com/>`_ for information on how to create your own container.
+
+
+Modifying the training script
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (akihironitta) [inline docs/source/tutorial/multi_node_multi_gpu_vanilla.rst:80]:
```suggestion
Modifying the training script
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by akihironitta — APPROVED** [review]
I'm not a slurm expert, but LGTM. Just pushed a few commits, but feel free to revert them if you think otherwise.

**Comment by flxmr** [comment]
I didn't find how to make the `srun`-only example work with the containers. The current example which I just copied doesn't work: there is no environment variables in the pytorch-only NGC-container and I wonder how those would be injected anyway. Also the mounting doesn't make sense. So now it is sbatch only, but verified to work on our local HPC (which also has a tutorial on modifying the enroot-container).

In addition I noticed by trying this, that having multiple processes download data and then trying to unzip it is not working well (it worked previously, because the data was downloaded already). Fixed that too.

**Review by puririshi98 — APPROVED** [review]
LGTM, thanks for the improvements :)

**Review by puririshi98 — CHANGES_REQUESTED** [review]
Although it looks good to me we should definitely make sure containers work as my original PR was tested working on a pyxis enabled nvidia cluster using our NVIDIA early access PyG container.

**Comment by flxmr** [comment]
So, I checked again and it seems `pyxis/enroot` is indeed injecting the `LOCAL_RANK` and `RANK` into the container: https://github.com/NVIDIA/enroot/blob/master/conf/hooks/extra/50-slurm-pytorch.sh#L32-L37

it seems to be quite a mess though (user→slurm→pyxis→enroot): https://github.com/NVIDIA/pyxis/issues/46#issuecomment-840283874

As for the `MASTER_ADDRESS` - maybe this is site-configuration specific (I could imagine running a [prolog](https://slurm.schedmd.com/prolog_epilog.html)), [they have it](http://projects.dfki.uni-kl.de/km-publications/web/ML/core/hpc-doc/docs/slurm-cluster/multi-gpu-distributed-jobs/) our HPC centers DGX don't have it!

**Comment by puririshi98** [comment]
@flxmr  I will investigate and get back to you as soon as I can

**Comment by puririshi98** [comment]
@flxmr i asked around internally with our enroot team 
```
Pytorch is supported through this enroot hook: https://github.com/NVIDIA/enroot/blob/master/conf/hooks/extra/50-slurm-pytorch.sh
It is shipped by default and just need to be enabled by the administrator
```
does this help? I can follow back up if not.

**Comment by flxmr** [comment]
So, hope this works now for everyone. I added a link to this issue because I suppose if our HPC didn't like the hook, others might not like it too.

**Review by puririshi98 — APPROVED** [review]
LGTM now, thanks @flxmr for this great PR

**Pull request merged**